Add Krivodonova hierarchical limiter

nilsdeppe opened the pull request:
## Proposed changes

**Note:** Only the last commit is new "Add Krivodonova Limiter"

Adds the hierarchical limiter of Krivodonova.

Depends on:
- [x] #1337 
- [x] #1352 in the sense that it assumes #1352 will be merged first, but I've added a fixup commit that can be squashed in to get around this.
- [x] #1376 (loosely)
- [x] #1377 
- [x] #1380 

Commit order:
- Add ModalTag prefix tag
- Add function to get collapsed index from Index
- Factor Mesh tests
- Add function to compute collapsed index in Mesh
- Add Krivodonova limiter
- add_krivodonova_limiter fixup: Add Krivodonova limiter

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add Krivodonova limiter

Adds the hierarchical slope limiter of Krivodonova that acts on the spectral
coefficients from the highest down, limiting them each individually.

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,969 @@
+
+std::ostream& operator<<(std::ostream& os, HierarchicalType t) noexcept;
+
+/// Given the `NodalTag` swap the `DataVector` with a `ModalVector`.
```
I'm not exactly sure where this tag should go. It's sort of a prefix tag, so that might be the place to put it.

fmahebert replied:
In Minmod I have the same kind of tag, except I swap the DataVector for a double... maybe we need to start collecting them.

nilsdeppe replied:
Hmm, yea, that's a good idea. Probably a special group of prefix tags that just swap the type

fmahebert replied:
I am (slowly) working on a refactor of Minmod, which will probably be merged after this PR. Shall I make the change in my upcoming PR, or do you want to do it here?

nilsdeppe replied:
Yes, I can factor this out. Probably the most reasonable place is `DataStructures/Tags.hpp`

nilsvu replied:
`DataStructures/Tags.hpp` sounds good 👍

nilsdeppe commented:
Did a review on my own to factor some parts of the code out a bit. Should be marginally easier to parse :)

nilsvu commented:
can review on Monday

nilsdeppe commented:
Rebased to see if Travis fix works

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+/// \endcond
+
+namespace SlopeLimiters {
+enum class HierarchicalType { Krivodonova };
```
Since you've thought about hierarchical limiters more than me: do you anticipate that other Hierarchical-type limiters will share enough code with the Krivodonova limiter that it makes sense to have them be switches on the same class (instead of one independent class per hierarchical limiter)?

nilsdeppe replied:
I'm on the fence with this. I like the idea of being able to switch between the limiters similarly to what is done in minmod, even though the Krivodonova limiter is the only one that requires parameters. Almost no code is shared other than the boiler plate for options.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+  };
+
+  /*!
+   * \brief The \f$\alpha_i\f$ values in the Krivodonova algorithm.
```
- Does it make sense to specify each `alpha_i` by hand? The Krivodonova paper conveniently gives a range `[alpha_{min,i}, alpha_{max,i}]` for each `alpha_i`, so a simpler approach (though slightly less general) would be to specify a single linear scaling parameter in `[0, 1]` to the limiter, which would control how far from `alpha_{min,i}` to `alpha_{max,i}` every `alpha_i` is picked. In practice, I don't know if we will need independent control on how each deriv order is limited, but if even we do, I worry that specifying each `alpha_i` by hand is prone to all sorts of errors (in 3D, this is a long list with a non-trivial ordering, and it is hard to know if the user gave every factor in the intended order). If the single-parameter way of specifying alpha that I am proposing is too simplistic, I think it may be better to have a flag (i.e. enum) to represent variations like "flat alpha", "bigger alpha for large coeffs", "smaller alpha for large coeffs", etc, without making the user specify tens of doubles in (hopefully) the right order...

- Do other hierarchical limiters also use this option? If not, this feeds into my concern that different hierarchical limiters may be best implemented as different classes (but again, I am not familiar with the other formulations of hierarchical limiters).

nilsdeppe replied:
- The ability to change the alphas is useful and even in 3D there are only 10 that can be specified since we are limited to 12 coefficients. The Krivodonova range comes from the idea of not forcing the higher derivatives to be lesser than or equal to the maximum finite difference of the lower derivatives times some scaling factor (the alphas). The lower bound comes from the Legendre polynomials (it'll be a slightly different factor for Chebyshev, I assume), but really the lower bound is zero (shouldn't change the sign of the derivatives).

- All of them can, but they really try to be parameter free. I'm unsure of whether or not to combine them all in one class. I like the idea of being able to control which is being used from the input file, though in the long run that's probably not necessary.

fmahebert replied:
First item: In my first pass, I had not correctly understood that the alpha are a 1-dimensional set even in 2d/3d simulations. This makes me less opposed to specifying all of them 👍 (though it still sounds annoying).

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+    ASSERT(j < mesh.extents()[1], "second coefficient index must be less than "
+                                      << mesh.extents()[1] << " but is " << j);
+    return i + mesh.extents()[0] * j;
+  }
```
I have sometimes wanted this functionality without going all the way to an `IndexIterator`. Could this functionality (going between collapsed index and N-dim index) be promoted to a member function of (or, a publicly-available free function bundled with) `Mesh`?

nilsdeppe replied:
Yes, I can factor this out, but I'll need to think a bit about where it should go :)

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+      alphas_(std::move(alphas)),
+      disable_for_debugging_(disable_for_debugging) {
+  if (alg::any_of(alphas_, [](const double t) noexcept {
+        return t > 1.0 or t <= 0.0;
```
Why not check the stricter bounds of Krivodonova Eq 23 (modulo verifying that her convention for Legendre polynomial normalization matches ours)?

nilsdeppe replied:
Because her bounds are stricter than necessary. The lower bound is derived as the "exact" value that "should" be chosen, but being larger or smaller makes the limiter less or more dissipative, respectively.

fmahebert replied:
Okay, can you make a note in the documentation that the alpha are only checked to lie in [0,1] which is a looser interval than the one recommended in the paper?

fmahebert replied:
Maybe here also add a note ~ "See the main documentation for an explanation of why these bounds are different from those of Krivodonova 2007"

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+  }
+
+  packaged_data->volume_data.initialize(mesh.number_of_grid_points());
+  // perform phys to spec transformation on each tensor component
```
I think we're more consistently using nodal/modal over phys/spec.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+        std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>, PackagedData,
+        boost::hash<std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>>>&
+        neighbor_data) const noexcept {
+  if (disable_for_debugging_) {
```
Depending on whether `UNLIKELY` was correct in previous function, this could also be `UNLIKELY`

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+    const gsl::not_null<PackagedData*> packaged_data,
+    const db::item_type<Tags>&... tensors, const Mesh<VolumeDim>& mesh,
+    const OrientationMap<VolumeDim>& orientation_map) const noexcept {
+  if (UNLIKELY(disable_for_debugging_)) {
```
For my education: Is this the correct use for `UNLIKELY`? I understood `UNLIKELY` to be for rare run-time branches, rather than input-file options that may be rarely specified but would be true 100% in those particular simulations...

nilsdeppe replied:
You're right that in the case of debugging it is 100% hit, but that is the `UNLIKELY` case (in production we won't have this flag be true)

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+          << hierarchical_type_
+          << "' does not yet support non-uniform number of collocation points "
+             "in each direction. The extents are: "
+          << mesh.extents());
```
- The error and its message are inconsistent. Are you intending to check only the number of points (as in the message), or the number of points and the type of basis and the type of quadrature (as in the logic)?

- I'm guessing you could go even stricter and require a Legendre basis; doesn't the Krivodonova paper assume Legendre when deriving the acceptable ratios between spectral coefs? However, I think either Gauss or GaussLobatto quadrature should be fine.

nilsvu replied:
ping, this still checks for extents, basis and quadrature but the error message only mentions the extents

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+    // for each coefficient...
+    for (size_t i = mesh.extents()[0] - 1; i > 0; --i) {
+      for (size_t j = i; j < mesh.extents()[1]; --j) {
+        // Check if we are done limiting
```
A more helpful comment would be to write something like `// We are done limiting this tensor component, move on to next component` above each of the two `goto`s.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+    for (size_t i = mesh.extents()[0] - 1; i > 0; --i) {
+      for (size_t j = i; j < mesh.extents()[1]; --j) {
+        for (size_t k = j; k < mesh.extents()[2]; --k) {
+          // Check if we are done limiting
```
Same comment as in 2D

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+          // Check if we are done limiting
+          auto& self_coeffs = get<Tag>(*coeffs_self)[tensor_storage_index];
+          // We treat the different cases separately to reduce the number of
+          // times we call minmod, not because it is required for correctness.
```
Good comment 👍

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+            self_coeffs[storage_index(mesh, i, j, k)] = tmp_ijk;
+            self_coeffs[storage_index(mesh, k, i, j)] = tmp_kij;
+            self_coeffs[storage_index(mesh, k, j, i)] = tmp_kji;
+          } else {
```
What about the case `} else if (i == k and i != j) {` ... is it redundant and I'm failing to see the logic, or is it missing?

nilsdeppe replied:
`i >= j >= k` so the case you describe cannot happen (I was confused by this for a bit too while writing initially :) )

fmahebert replied:
Oh, I see it now. Could you add a (brief) comment to this effect?

nilsdeppe replied:
Yes, good idea :)

fmahebert reviewed (changes requested):
A quick first pass. Haven't yet looked in detail at the documentation or the test.

nilsdeppe commented:
@fmahebert pushed commits to address your comments. I factored out some of the changes into new commits because they didn't make sense in the `Add Krivodonova limiter` commit. I also rebased on master since #1337 got merged.

nilsdeppe commented:
I missed a function for `Index<0>` which caused linking errors.

nilsvu commented on docs/References.bib:
```diff
@@ -248,6 +248,20 @@ @article{Kozlowski1978aa
+  pages =        "879 - 896",
+  year =         2007,
+  issn =         "0021-9991",
+  doi =          "https://doi.org/10.1016/j.jcp.2007.05.011",
```
- Make this the `url` and add a separate `doi` field without the `doi.org`.
- Align `=` like the rest of the file
- Enclose all values in `""`
- Format author as `Krivodonova, Lilia`
- You don't need the keywords, probably better remove them

nilsdeppe replied:
I don't see why the `=` location is important. It's not my fault you aligned them incorrectly ;) I'm happy to reformat the entire file if you prefer :)

nilsvu replied:
It's for consistency with inspire-hep

nilsvu commented on src/DataStructures/Tags.hpp:
```diff
@@ -0,0 +1,25 @@
+namespace Tags {
+/// Given the `NodalTag` swap the `Tensor::type` with a `ModalVector`.
+template <typename NodalTag>
+struct ModalTag : db::SimpleTag {
```
Better remove the `Tag` from the name I think, it's redundant

nilsvu commented on src/DataStructures/Tags.hpp:
```diff
@@ -0,0 +1,25 @@
+  using type =
+      TensorMetafunctions::swap_type<ModalVector, typename NodalTag::type>;
+  static std::string name() noexcept {
+    return "coefficients(" + NodalTag::name() + ")";
```
nodal data are also coefficients (to Lagrange polynomials). So better write e.g. `"Modes(...)"`

nilsvu commented on tests/Unit/DataStructures/Test_Tags.cpp:
```diff
@@ -0,0 +1,51 @@
+  static_assert(cpp17::is_same_v<typename Tags::ModalTag<VectorTag<3>>::type,
+                                 tnsr::I<ModalVector, 3>>,
+                "Failed testing tags::ModalTag");
+  CHECK(Tags::ModalTag<ScalarTag>::name() == "coefficients(Scalar)");
```
Does the `Modal` tag work in as a prefix for `Tags::Variables`? Would be good to add a test for that

nilsdeppe replied:
No, I don't think it does, nor am I convinced it should.

nilsvu replied:
why not? we can swap tensor dtypes, and a Variables is just a collection of tensors?

nilsvu commented on src/DataStructures/Index.hpp:
```diff
@@ -77,6 +77,12 @@ class Index {
+   * \brief Returns the 1-dimensional index corresponding to the `Dim`
+   * dimensional `index`.
+   */
+  size_t collapsed_index(const std::array<size_t, Dim>& index) const noexcept;
```
- Can you not use the existing function? If you can't, please add comments why we need both
- Can you move the implementation to the cpp file?

nilsdeppe replied:
- What existing function?
- No, why would I do that? I want this inlined.

nilsvu replied:
The function is currently in this same file, with the same name, but in the global namespace

nilsdeppe replied:
I've added a fixup that deals with this. It should be resolved there.

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+ *
+ * Implements hierarchical slope limiters that limit the logical derivatives of
+ * the `Tags` starting at the highest derivatives. The simplest of these
+ * limiters is described in Krivodonova 2007 \cite Krivodonova2007. The
```
I think better remove author/year, it is available through the `\cite`

nilsdeppe replied:
No, she has multiple papers on this subject.

nilsvu replied:
In the documentation the bibliographical information is available on-hover, so I'd prefer to reduce redundancy, e.g. `is described in \cite Krivodonova2007`. That's the same information in the source code, and it's nicely rendered in the HTML.

nilsdeppe replied:
I still don't agree with you

nilsdeppe replied:
okay, I've done this, it's in a new fixup

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+ * Implements hierarchical slope limiters that limit the logical derivatives of
+ * the `Tags` starting at the highest derivatives. The simplest of these
+ * limiters is described in Krivodonova 2007 \cite Krivodonova2007. The
+ * Krivodonova limiter is applied to the spectral coefficients. In 3d the
```
`3D`?

nilsdeppe replied:
Nope, `3d`.

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+ * function being limited is expanded as:
+ *
+ * \f{align}{
+ * u^{l,m,n}=\sum_{i,j,k=0,0,0}^{N_i,N_j,N_k}c^{l,m,n}_{i,j,k}
```
remove the `,` from the indices (except you have a reason for those)

nilsdeppe replied:
Nope, I want them there, and you didn't say which ones.

nilsvu replied:
We don't commonly use `,` to separate indices in physics, in particular for relativists these read more like partial derivatives. This is in the context of the Krivodonova paper and she uses that notation, so I don't feel strongly about this.

nilsdeppe replied:
"We" don't normally have indices that have the same meaning as the upper indices here. I understand your argument for the lower ones. I'm not still not sure which you are objecting to :)

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+ * \f}
+ *
+ * where \f$\left\{\xi, \eta, \zeta\right\}\f$ are the logical coordinates,
+ * \f$P_{i}\f$ the basis functions (typically Legendre polynomials), and the
```
In `Spectral` the basis functions are denoted with `\phi`

nilsdeppe replied:
So? They're the Legendre polynomials, which are `P`.

nilsdeppe replied:
And in Spectral they appear to be incorrectly denoted by `\Phi` instead of `\phi`. The capital ones are reserved for spacetime test functions, not spatial ones.

nilsvu replied:
you say "typically Legendre polynomials". If this is general, I think `\phi` should be used. If Legendre polynomials are indeed assumed here, then `P`.

nilsdeppe replied:
The original derivation and the lower bound by Krivodonova are for Legendre polynomials. However, the procedure is general for any 1d basis (could probably be generalized to 2d bases but :man_shrugging: ). I will change to `\phi` and reword to make it clear that Krivodonova used Legendre, but that the procedure is general.

nilsdeppe replied:
I'm leaving it as `P` focusing on Legendre polynomials and mentioning where the fact that Legendre polynomials are needed that we relax this requirement and how (the lower bound of the alphas)

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+ *   &\left(c_{0,0,1}^{l,m,n},
+ *          \alpha_1\left(c^{l,m,n+1}_{0,0,0}-c^{l,m,n}_{0,0,0}\right),
+ *          \alpha_1\left(c^{l,m,n}_{0,0,0}-c^{l,m,n-1}_{0,0,0}\right)
+ *     \right),
```
(a) did you type all of this? 😄 and (b) would a shorter example suffice, perhaps in 2d?

nilsdeppe replied:
- yes
- no, I found the 2d example insufficient.

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+ *   coefficients. In the case that the neighbor has more coefficients we can
+ *   just ignore the higher coefficients. In the case that the neighbor has
+ *   fewer coefficients we have a few choices that will probably depend on which
+ *   specific HL is used.
```
don't abbreviate `HL`

nilsdeppe replied:
Why?

nilsvu replied:
It's the first occurrence of that abbreviation. I guess it means hierarchical limiter, which is not a common term by any means. If you must use it, please introduce it at some point.

nilsdeppe replied:
Okay, I will introduce the abbreviation at the beginning :+1:

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+ *   appropriately again.
+ * - h-refinement is not supported, but is straightforward to add since the
+ *   differences in coefficients in Eq.\f$\ref{eq:krivodonova 3d minmod}\f$ will
+ *   just be multiplied by non-unity factors.
```
It seems like you comment on a number of things here that are straightforward to add. Which will we actually need in the near future? For the shock tests, support for p-anisotropic meshes would be useful. With AMR coming up, h-refinement will be important. Are you planning to add these features in upcoming PRs?

nilsdeppe replied:
All this depends on how well the limiter works. I don't see why we need anisotropic p for the shock tests. I'd argue that's actually the wrong thing to test.

nilsdeppe replied:
I am planning to add these in the future, but how soon will depend on how well the limiter works on the disk. If it doesn't perform well then it'll be a while out, if it performs well it'll be sooner.

nilsvu replied:
👍 on the extended mesh checks

fmahebert replied:
It is not at all obvious to me that the h-refinement case is straightfoward. With multiple neighbors across a face, it is not obvious how the relationships between the spectral coeffs of neighboring cells are affected. If you have it all worked out, that's fine, but otherwise I wouldn't claim too much here.

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+   */
+  struct Alphas {
+    using type = std::vector<double>;
+    static constexpr OptionString help = {"The alpha coefficients"};
```
"coefficients" is used for nodal and modal coefficients a lot. Perhaps "parameters", or use the help string to describe these briefly. I prefer to have the option name and help string not be redundant ("Alphas: The alpha parameters" doesn't really add much information)

nilsdeppe replied:
I can call it the `rescaling parameters` but that also doesn't help. Basically, anyone will need to read the actual documentation regardless.

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+    static constexpr OptionString help = {"The alpha coefficients"};
+    static type default_value() noexcept {
+      return std::vector<double>(
+          Spectral::maximum_number_of_points<Spectral::Basis::Legendre>, 1.0);
```
Please comment on the default `Legendre` polynomials in the documentation

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+      HierarchicalType hierarchical_type,
+      std::vector<double> alphas = std::vector<double>(
+          Spectral::maximum_number_of_points<Spectral::Basis::Legendre>, 1.0),
+      bool disable_for_debugging = false, const OptionContext& contex = {});
```
`context`

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,982 @@
+      bool disable_for_debugging = false, const OptionContext& contex = {});
+
+  Hierarchical() = default;
+  Hierarchical(const Hierarchical&) = default;
```
do you need the copy constructor?

nilsdeppe replied:
I did because of a DataBox perfect forwarding bug, I'm adding a commit that fixes that so copy semantics won't be needed.

nilsvu reviewed (changes requested):
haven't read the tests yet

nilsdeppe commented:
@nilsleiffischer fixups pushed, added a commit that fixes the References.bib file

nilsdeppe commented:
@nilsleiffischer I think I've either responded to or fixed all your comments

nilsdeppe commented:
@fmahebert @nilsleiffischer I'm going to factor out all commits except "Add Krivodonova limiter". If you'd like to give feedback on the fixups for those then please do so today. I'm going to do the factoring tonight. Thanks!

nilsvu commented on docs/References.bib:
```diff
@@ -248,6 +249,19 @@ @article{Kozlowski1978aa
 }
 
+@article{Krivodonova2007,
+  title   = "Limiters for high-order discontinuous Galerkin methods",
```
Add `{}` braces around `Galerkin` to preserve capitalisation

nilsvu commented on docs/References.bib:
```diff
@@ -97,9 +97,9 @@ @article{Fishbone1976apj
-                        Scheel, Mark A. and Szilagyi, Bela",
+                        and Duez, Matthew D. and Haas, Roland and Kidder,
+                        Lawrence E. and Ott, Christian D. and Pfeiffer,
+                        Harald P. and Scheel, Mark A. and Szilagyi, Bela",
```
Since you're touching this, please fix the indentation of the entry to 2 spaces like the rest of the file. (we really have to add automated checks for these things to keep this file tidy and consistent...)

nilsdeppe replied:
Until there's a check for that I'm not changing it.

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -432,10 +432,14 @@ class Hierarchical<VolumeDim, tmpl::list<Tags...>> {
     using type = std::vector<double>;
-    static constexpr OptionString help = {"The alpha coefficients"};
+    static constexpr OptionString help = {
+        "The alpha parameters of Krivodonova limiter"};
```
How about `The alpha parameters by Krivodonova`? (nitpicking here.. trying to get rid of redundancy in option descriptions)

nilsdeppe replied:
I don't like that phrasing, personally. I've updated to `The alpha parameters of the Krivodonova limiter`

nilsvu reviewed (changes requested):
I'm fine with squashing and refactoring 👍

fmahebert commented:
I too am happy with squashing the current fixups and factoring 👍

nilsdeppe commented:
Squashed, rebased on develop, and add PRs factoring out the commits. Also added a fixup for the remaining open comments.

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2383 @@
+                  const std::vector<double>& coeffs) noexcept {
+  (*result) = 0.0;
+  DataVector temp(result->size());
+  for (IndexIterator<Dim> it(mesh.extents()); it; ++it) {
```
Add short comment similar to `// Iterator ranges over spectral modes`, or alternatively just rename `it` to `mode_iter` or similar.

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2383 @@
+//
+// where the coefficients c_{i,j,k} are the 1D `coeffs`.
+template <size_t Dim>
+void set_legendre(const gsl::not_null<DataVector*> result,
```
Please use a more descriptive name. I suggest `assemble_from_legendre_modes`, but probably we can do better than that.

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2383 @@
+    std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>, PackagedData,
+    boost::hash<std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>>>;
+
+template <size_t I>
```
What is this template?

nilsdeppe replied:
I'm using it to be able to have 2 `Scalar` tags, `0`, and `1`. Maybe `Identifier` would be a better name?

fmahebert replied:
Yes, I would like `Identifier` much better

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2383 @@
+  using type = Scalar<DataVector>;
+};
+
+template <size_t Dim, size_t I>
```
What is the second template?

nilsdeppe replied:
Let's continue this conversation on the Scalar tag :)

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,991 @@
+  bool operator==(const Hierarchical& rhs) const noexcept;
+
+  struct PackagedData {
+    Variables<tmpl::list<::Tags::Modal<Tags>...>> volume_data;
```
I think it would make it even clearer that this is modal data if the variable name said it too... how about `volume_modal_data`? I'm open to other better ideas too :)

nilsdeppe replied:
I like `modal_volume_data` but :man_shrugging: :) I generally like the idea of adding `modal` to the name

fmahebert replied:
That works too :)

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2383 @@
+                                                                        x) +
+      2.0 * Spectral::compute_basis_function_value<Spectral::Basis::Legendre>(
+                order - 1, x);
+  Limiter::PackagedData package_data0{};
```
`package_data0` -> `packaged_data`

The `0` seems unnecessary. Using past tense will make this appear less like a function name.

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2383 @@
+
+  hierarchical.package_data(make_not_null(&package_data0), tensor0, tensor1,
+                            mesh, {});
+  ModalVector result_tensor0(mesh.number_of_grid_points(), 0.0);
```
`result_tensor0` -> `expected0` (Or make it actually be a tensor)

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2383 @@
+  result_tensor0[mesh.number_of_grid_points() - 1] = 1.0;
+  CHECK(get(get<::Tags::Modal<ScalarTag<0>>>(package_data0.volume_data)) ==
+        result_tensor0);
+  ModalVector result_tensor1(mesh.number_of_grid_points(), 0.0);
```
`expected1` (Or make it actually be a tensor)

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2383 @@
+  package_data_lower.volume_data.initialize(num_pts);
+  package_data_lower.mesh = mesh;
+
+  Scalar<DataVector> self_lnc(num_pts, 0.0);
```
Use a more helpful name than "lnc", this isn't f77. I suggest: `data_to_limit` or similar, but am happy to consider alternatives.

nilsdeppe replied:
Buuutttttttt...... :stuck_out_tongue: `lnc` "obviously" stands for `limit no coefficient` (a historic relic of an early draft of these tests)

nilsdeppe replied:
`nodal_scalar_data_to_limit`?

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2383 @@
+      CHECK_ITERABLE_APPROX(self_vector.get(i), expected);
+    }
+  };
+  // Note: Throughout the tests 1.0e-18 is used as a +0.
```
Explain why you do this. (Is it to obtain fine control over which branches the limiting procedure takes?)

nilsdeppe replied:
Yes, I've added a comment :)

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2383 @@
+          .get(i) = upper_coeffs;
+      get<::Tags::Modal<VectorTag<dim, 0>>>(package_data_lower.volume_data)
+          .get(i) = lower_coeffs;
+    }
```
I think it would be a much more interesting test to have a _single_ limiter call where the Scalar and Vector inputs have different values and therefore different logical paths. Here you make a large number of calls but because all the variable components are identical, the test can't find bugs mixing different tensor components (obviously unlikely).

I would suggest cleaning the 1D tests you have now to use just the Scalar, then add a single limiter invocation in which you test limiting both the Scalar and Vector with more non-identical inputs.

nilsdeppe replied:
I don't disagree that testing a call of the different limiters is helpful to test that the components are not mixed. The current test (which I'd like to keep) makes sure that both Scalar's and Tensor's have all their components limited. I think testing a single case for the limiter not mixing tensors/components and testing that in all cases the limiter is applied to all components is best.

I'll add the suggested test case :)

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2383 @@
+void run() noexcept {
+  for (size_t order = 2; order < 7; ++order) {
+    package_data(order);
+  }
```
Testing `package_data` at a whole bunch of different orders is really just retesting our spectral transforms, so I don't think it's necessary. More important, I think, would be testing that the different limiter inputs are correctly orientated, especially in 2 and 3 dimensions.

I suggest: just test this at one order, but make sure to test at least one case per dimension with a non-identity orientation.

nilsdeppe replied:
Agreed, I think in 1d testing the different orders is not bad, but I don't think it's necessary for all dims. I will add tests for the different orientations and higher dims.

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2383 @@
+  // Note: Throughout the tests 1.0e-18 is used as a +0.
+
+  // Limit no coefficients because the highest coefficient doesn't need
+  // limiting.
```
Please add a comment giving the thresholds (i.e., what is the magical value that you are tweaking coeffs to be above/below, and how is this magical value optained?), so the test is easier to parse and/or modify in the future.

nilsdeppe replied:
There is no magic value, it's computed by the minmod function from the finite difference. It's given in detail in the documentation for a 3d element. I'm not sure it would be enlightening to copy-paste the documentation here.

fmahebert replied:
I agree, looking more carefully this seems clear enough, and more importantly, making it any clearer would indeed require nearly duplicating the docs...

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2383 @@
+         {3.0, -2.0, 2.0, 0.0});
+
+  // Limit the top 2 coeffs:
+  // Limit top upper neighbor, limit second upper neighbor
```
When getting to second coeff, no need to explain how the threshold is obtained, but it would be nice to state its value anyway.

nilsdeppe replied:
Sorry, I'm not understanding what you're asking for here. Could you rephrase and/or give an example?

fmahebert replied:
As prior comment, I now think this is clear enough (and that it would be a large effort to make it clearer)

fmahebert reviewed (changes requested):
Some comments from starting to look at the tests (not done yet)

nilsdeppe committed:
Allow larger file sizes

nilsdeppe commented:
@fmahebert pushed the tests you requested as fixups :)

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+ *
+ * Implements HL's that limit the logical derivatives of the `Tags` starting at
+ * the highest derivatives. The simplest of these limiters is described in \cite
+ * Krivodonova2007. The Krivodonova limiter is applied to the spectral
```
When I render the docs, this citation doesn't link to anything, and just renders as (in bold:) **[Krivodonova2007]**.

edit: I haven't updated my container in a while so perhaps that's on me?

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+ *  \begin{array}{ll}
+ *    \mathrm{sgn}(a)\min(\lvert a\rvert, \lvert b\rvert,
+ *    \lvert c\rvert) & \mathrm{if} \;
+ *    \mathrm{sgn}(a)=\mathrm{sgn}(b)=\mathrm{sgn}(c) \\ 0 & \mathrm{otherwise}
```
Here you define the three-argument minmod (which is the common one), but you are using a minmod with many more arguments in the surrounding documentation. Perhaps most easily resolved by writing "minmod(a, b, c, ...)" here?

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+ *
+ * \f{align*}{
+ * \frac{1}{2(2i-1)}\le \alpha_i \le 1
+ * \f}
```
Krivodonova gives this range in 1D, but a different range in higher dimensions, because she changes the normalization of the Legendre polynomials halfway through the paper. Please check how Spectre's normalization compares to Krivodonova's normalization, and either:
- update this equation to match Krivodonova Eq 23, if we match her 3D normalization, OR
- add a note that this differs from her Eq 23 because we do not match her 3D normalization.

nilsdeppe replied:
She doesn't do anything in 3D, she only does 2D. Maybe I'm  confused about this, but our Legendre polynomials are always normalized to P(1)=1 independent of dimension, so would that change the case in a tensor product basis?

fmahebert replied:
Krivodonova normalizes her 2D (you're right, there is no 3D) Legendre polynomials so that `integral P_i P_j = delta_ij`. This is a different normalization, and leads to a different set of bounds on alpha. As a result, your documentation does not obviously match the paper when dim > 1, and you need to document this difference, so the reader can usefully compare your code to the paper. Add a note saying, roughly:

"Note that we normalize our Legendre polynomials by P(1) = 1; this is the normalization Krivodonova uses in 1D, (but not in 2D), which is why our bounds on alpha match Krivodonova Eq. 14 (but not Eq. 23)."

nilsdeppe replied:
Okay, done.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+ * \frac{1}{2(2i-1)}\le \alpha_i \le 1
+ * \f}
+ *
+ * where the lower bound comes from the exact finite differencing when using
```
"comes from the exact finite differencing" doesn't mean anything to me. I can't suggest a replacement because I don't know what you are trying to say. The differencing of what? Why does that lead to a bound?

nilsdeppe replied:
I'll add a parenthetical that one should look at the paper for details. The lower bound is the value you get from doing a finite difference to approximate the higher order derivative from the lower derivatives.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+ * 0 \le \alpha_i \le 1
+ * \f}
+ *
+ * to allow different polynomials and to allow the limiter to be more
```
"different polynomials" ... different from what?

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+ * dissipative if necessary. The same \f$\alpha_i\f$s are used in all
+ * dimensions. The only place where the specific choice of 1d basis comes in is
+ * the lower bound for the \f$\alpha_i\f$s, and so in general the limiter can be
+ * applied to any 1d or tensor product of 1d basis functions.
```
This last sentence seems like it could be a \note or somehow less prominent.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+ * element with \f$N+1\f$ coefficients in each dimension and denote the
+ * coefficients as \f$c_{i,j,k}\f$. Then the limiting procedure starts at
+ * \f$c_{N,N,N}\f$, followed by \f$c_{N,N,N-1}\f$, \f$c_{N,N-1,N}\f$, and
+ * \f$c_{N-1,N,N}\f$, etc. Limiting is stopped if all symmetric pairs of
```
Give more cases before falling back to "etc"... the reader should know: after the permutations of {N,N,N-1} will there be permutations of {N,N,N-2} or {N,N-1,N-1}? Or rephrase to make clear you will address this further down in the example.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+ * applied to any 1d or tensor product of 1d basis functions.
+ *
+ * The limiting procedure must be applied from the highest derivatives to the
+ * lowest, or the highest coefficients to the lowest. Let us consider a 3d
```
"or" -> "i.e.," (I assume you are clarifying, not giving an alternative)

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+ * coefficients as \f$c_{i,j,k}\f$. Then the limiting procedure starts at
+ * \f$c_{N,N,N}\f$, followed by \f$c_{N,N,N-1}\f$, \f$c_{N,N-1,N}\f$, and
+ * \f$c_{N-1,N,N}\f$, etc. Limiting is stopped if all symmetric pairs of
+ * coefficients are not limited. That is, if \f$c_{N-i,N-j,N-k}\f$,
```
You need to define what "are not limited" means in terms of c vs. tilde{c} probably.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+ * \f$c_{N-j,N-i,N-k}\f$,  \f$c_{N-k,N-j,N-i}\f$, \f$c_{N-j,N-k,N-i}\f$,
+ * \f$c_{N-i,N-k,N-j}\f$, and \f$c_{N-k,N-i,N-j}\f$ are not limited, we have
+ * finished limiting. As a concrete example, consider a 3d element with 3
+ * collocation points per dimension. The coefficients are adjusted according to:
```
In this example there is some lack of clarity w.r.t. order of operations. It sounds like all the tilde{c}s are computed first, but later (below all the equations) you tell the reader how the stopping condition works.

I would rephrase this sentence here to say "Each limited coefficient is defined by:"

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+ * \f}
+ *
+ *
+ * Finally, limiting is performed as follows:
```
(...continued from comment above equations)

And here I would say "The algorithm to perform the limiting is as follows"

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+ *
+ * Finally, limiting is performed as follows:
+ *
+ * - limit \f$c_{2,2,2}\f$, if not changed, stop
```
Here again it would be good to clarify how the words "limit", "changed" are related to the distinction of c vs tilde{c}.

I suggest "limit c (i.e., c <- tilde{c}); if the value is unchanged, then stop" in the first line, then in all the algorithm steps below you do not need to restate because it will already be clear.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+ *
+ * #### Limitations:
+ *
+ * - We are recomputing the spectral coefficients after sending the coefficients
```
I think this can be slightly clearer if rephrase as "We currently recompute the spectral coefficients for local use, after having computed these same coefficients for sending to the neighbors"

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+ *   function to return an object to be passed as an additional argument to the
+ *   `operator()` (still would need to be stored in the DataBox).
+ * - We cannot handle the case where neighbors have more/fewer
+ *   coefficients. In the case that the neighbor has more coefficients we can
```
can -> could, because we don't handle it yet :P

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+ *   just ignore the higher coefficients. In the case that the neighbor has
+ *   fewer coefficients we have a few choices that will probably depend on which
+ *   specific HL is used.
+ * - Having different number of collocation points in different directions is
```
"a different number" or "different numbers"

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+ *   specific HL is used.
+ * - Having different number of collocation points in different directions is
+ *   not supported. However, it is straightforward to handle this case. The
+ *   outer most loop is in the direction with the most collocation points, while
```
- single word: outermost
- "is in the direction" -> "should be in the direction" (because as far as I know you don't handle this case yet)

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+ * - Having different number of collocation points in different directions is
+ *   not supported. However, it is straightforward to handle this case. The
+ *   outer most loop is in the direction with the most collocation points, while
+ *   the inner most loop is over the direction with the fewest collocation
```
again, "is" -> "should be"

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,998 @@
+   * \brief The \f$\alpha_i\f$ values in the Krivodonova algorithm.
+   *
+   * The default size is the number of allowed Legendre polynomials
+   * (`Spectral::maximum_number_of_points<Spectral::Basis::Legendre`)
```
- "is the maximum number of allowed Legendre basis polynomials"
- missing `>` in code snippet

fmahebert reviewed (changes requested):
After passing through the documentation, I feel more strongly that this should be a Krivodonova-only limiter. Mostly I feel this way because very little of the documentation or the implementation seems to actually be general, the Krivodonova "alpha" option is decidedly not general, and calling it a general implementation doesn't make it be one... 

If you feel strongly about a single entry point (I relate to the pain of recompilation), I would suggest we either shorten the documentation dramatically or find a way of splitting it up, then split the main source code into various files, so that the "general implementation" can be a thin wrapper that delegates to the correct HL. Or perhaps you have other ideas for ways to keep this general but readable...

nilsdeppe commented:
I agree with making it a specific implementation. I'll address the rest of the comments first and once everything is squashed I'll do the rename. I think that'll make it easiest to review.

nilsdeppe commented:
@fmahebert pushed fixups and rebased on develop. I'll do the rename once you're happy with the fixups and I squash them.

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,1044 @@
+ * \ingroup SlopeLimitersGroup
+ * \brief An implementation of the Krivodonova slope limiter.
+ *
+ * Implements HL's that limit the logical derivatives of the `Tags` starting at
```
spell out the abbreviation "HL" at this first occurrence. If you want to use this abbreviation later, you can introduce it here (however I'd prefer to just spell it out, it's not obvious what it means, especially when you rename the type to Krivodonova. I guess it means hierarchical limiter)

fmahebert replied:
This will probably change as part of the rename.

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,1044 @@
+      "A collection of hierarchical slope limiters.\n\n"
+      "Currently only the Krivodonova limiter is implemented.\n\n"
+      "Hierarchical slope limiters work by limiting the highest "
+      "coefficients/derivatives using an aggressive approach such as minmod, "
```
`modal coefficients`

fmahebert replied:
Agreed. Also note this string will have to change after the rename.

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,1044 @@
+    return hierarchical_type_;
+  }
+
+  bool operator==(const Hierarchical& rhs) const noexcept;
```
did I miss the `!=` operator?

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,1044 @@
+      if (UNLIKELY(id_packaged_data.second.mesh.basis(d) != mesh.basis(d))) {
+        ERROR(
+            "Must have the same basis in all dimensions but is not the "
+            "same in dimension: "
```
"Neighbours must have the same basis" (same below for quadrature)

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,1044 @@
+  expand_pack(fill_variables_tag_with_spectral_coeffs<Tags>(&coeffs_self,
+                                                            *tensors, mesh)...);
+
+  bool limited_any_component = false;
```
Add comment like `Perform the limiting on the modes` above this line

nilsdeppe replied:
I'm doing `modes` -> `modal coefficients` since I'm sure that change will be requested ;)

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,1044 @@
+}
+
+template <size_t VolumeDim, typename... Tags>
+bool operator==(
```
this should probably be `!=`

nilsvu commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2946 @@
+
+void run() noexcept {
+  INFO("Testing 1d limiter");
+  for (size_t order = 2; order < 7; ++order) {
```
you can use the `Spectral::minimum_number_of_points` and `Spectral::maximum_number_of_points` here

nilsvu commented on tools/Hooks/CheckFileSize.py:
```diff
@@ -8,22 +8,24 @@
 
 # The maximum file-size for a file to be committed in kB
-max_file_size = 130.0
+max_file_size = 200.0
```
😄

nilsvu reviewed (changes requested):
I'm fine with squashing so that you can do the name change

moxcodes commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2383 @@
+void test_2_2(const F& helper) noexcept {
+  // Limit no coefficients because the highest coefficient doesn't need
+  // limiting.
+  helper({0.0, 1.0, 2.0, 3.0, 4.0, 7.0, 3.0, 4.1, 0.0},
```
I understand that limiters are complicated, and it's important to hit every branch, but a lot of these long inputs seem to be very similar values with a couple of small alterations to make sure the limiter is applied to certain combinations. Is there no way that these values could be produced algorithmically? It causes me anxiety to consider the possibility of someone having to update these tests if something changes.

I also understand that finding an automatic way of producing these values could be a nontrivial undertaking, so I'm happy to just make an issue for this if you agree it is something that can and should be done.

nilsdeppe replied:
In theory, it's probably possible, in practice it's rather difficult because there are non-trivial dependencies between all the modal coefficients. My other concern with writing an algorithm is that that algorithm is basically the inverse of the limiter and given how complicated the limiter is I don't think it's any more likely that someone will implement the inverse algorithm correctly. I think testing individual numbers is the safest thing to do.

moxcodes replied:
Understood :+1: .

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,1044 @@
+  /*!
+   * \brief The \f$\alpha_i\f$ values in the Krivodonova algorithm.
+   *
+   * The default size is the maximum number of allowed Legendre basis
```
Perhaps document that it is okay to specify more alphas than there are grid points in any direction (if I understood that right). Do you check that enough alphas were given? In other words suppose I am using 4 points per dimension but give 3 alpha values... then what happens?

nilsdeppe replied:
I'm adding a parse error checking that we have as many as the max number of legendre grid points.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp:
```diff
@@ -0,0 +1,1044 @@
+            "the same in dimension: "
+            << d << ". Mesh is: " << mesh
+            << " neighbor mesh is: " << id_packaged_data.second.mesh);
+      }
```
Why do you check the neighbor mesh's extents, quadrature, and basis all separately? It seems that you could just demand the neighbor mesh be the same as the local mesh. Since you can print out the full mesh on an error, I think that comparing meshes directly should be sufficient.

nilsdeppe replied:
Somebody didn't want that before... That's what I did and @nilsleiffischer complains.

nilsdeppe replied:
Okay, I'm doing this. If there is disagreement, I'll let the reviewers come to a consensus and then adjust :)

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2946 @@
+//
+// where the coefficients c_{i,j,k} are the 1D `coeffs`.
+template <size_t Dim>
+void assemble_from_legendre_modal_coefficients(
```
Actually.. is this different from `to_nodal`? It looks like you always call this with `x` being the collocation points, in which case I think this should be identical to `to_nodal`...

nilsdeppe replied:
It appears to not be. I will do the replacement.

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2946 @@
+    for (size_t i = 0; i < dim; ++i) {
+      CHECK_ITERABLE_APPROX(nodal_vector_data_to_limit.get(i), expected);
+    }
+  };
```
Add blank line between lambda and comment

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2946 @@
+  // Note: Throughout the tests 1.0e-18 is used as a +0. since when we need the
+  // sgn of the coefficient to be +1, but the values to get the right results
+  // were done assuming zero. Since 1.0e-18 is positive but below machine
+  // precision for O(1) numbers it acts as +0.
```
This is clearer now 👍, but I think it can be more concise as follows:
"Note: Throughout the tests 1.0e-18 is used anywhere that we need a positive but zero coefficient, because it is below machine precision for O(1) numbers."

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp:
```diff
@@ -0,0 +1,2946 @@
+      ModalVector{7.0, 3.0, 2.5, 0.0};
+  get(get<::Tags::Modal<ScalarTag<0>>>(package_data_lo_xi.modal_volume_data)) =
+      ModalVector{1.0, -5.0, 1.0e-18, 0.0};
+  assemble_from_legendre_modal_coefficients(&get(nodal_scalar_expected), mesh,
```
Can you add a note (as you had in all the individual cases) saying which trigger this is? If I understand right, this is the top coef limited because of the upper neighbor?

(also add similar note for vector tag below)

fmahebert reviewed (commented):
I am happy with squashing, rebasing, and renaming.

nilsdeppe commented:
@nilsleiffischer @fmahebert I've rebased and squashed :)

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Krivodonova.hpp:
```diff
@@ -0,0 +1,1002 @@
+  /*!
+   * \brief The \f$\alpha_i\f$ values in the Krivodonova algorithm.
+   *
+   * The default size is the maximum number of allowed Legendre basis
```
"The default is 1. for all \f$\alpha_i\f$" and add that it always needs the maximum number of alphas

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Krivodonova.hpp:
```diff
@@ -0,0 +1,1002 @@
+ * \brief An implementation of the Krivodonova slope limiter.
+ *
+ * The slope limiter is described in \cite Krivodonova2007. The Krivodonova
+ * limiter is applied to the spectral coefficients. In 3d the function being
```
I think it would be nice to have a brief intuitive statement before diving into the technicalities here. I suggest (but up to you) using the nice OptionString text you wrote for the class and working it in here too.

Here's a possible way of doing this:

The slope limiter is described in \cite Krivodonova2007. The Krivodonova limiter works by limiting the highest derivatives/modal coefficients using an aggressive minmod approach, decreasing in derivative/modal coefficient order until no more limiting is necessary. In 3d, the function being limited is expanded as:

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Krivodonova.hpp:
```diff
@@ -0,0 +1,1002 @@
+ *
+ * \note The only place where the specific choice of 1d basis
+ * comes in is the lower bound for the \f$\alpha_i\f$s, and so in general the
+ * limiter can be applied to any 1d or tensor product of 1d basis functions.
```
This is a cool note, I hadn't appreciated this about the Krivodonova limiter before 👍

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Krivodonova.hpp:
```diff
@@ -0,0 +1,1002 @@
+   * (`Spectral::maximum_number_of_points<Spectral::Basis::Legendre>`)
+   */
+  struct Alphas {
+    using type = std::vector<double>;
```
Since now you expect a full list of 12, could this be a std::array?

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Krivodonova.hpp:
```diff
@@ -0,0 +1,1002 @@
+  static constexpr OptionString help = {
+      "The hierarchical slope limiter of Krivodonova.\n\n"
+      "This slope limiter works by limiting the highest "
+      "coefficients/derivatives using an aggressive minmod approach, "
```
modal coefficients, here :P

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Krivodonova.hpp:
```diff
@@ -0,0 +1,1002 @@
+                   Spectral::Basis::
+                       Legendre> << " number of alpha coefficients but got "
+            << alphas_.size());
+  }
```
I think this whole if statement can be avoided if you use a std::array?

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Krivodonova.hpp:
```diff
@@ -0,0 +1,1002 @@
+        "The Krivodonova limiter does not yet support non-uniform number of "
+        "collocation points, bases, and quadrature in each direction. The "
+        "extents are: "
+        << mesh.extents());
```
- "and quadrature" -> "or quadrature"
- could you print the whole mesh, instead of only its extents? then this error message will equally apply to a mesh with different extents but same quadratures, or same extents but different quadratures...

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Krivodonova.hpp:
```diff
@@ -0,0 +1,1002 @@
+    ERROR("The Krivodonova limiter does not yet support h-refinement");
+  }
+  alg::for_each(neighbor_data, [&mesh](const auto& id_packaged_data) noexcept {
+    for (size_t d = 0; d < VolumeDim; ++d) {
```
I think you shouldn't need this for loop? Can you just stick the `if` in the `for_each` ?

nilsdeppe replied:
This is quite possibly the strangest/dumbest line of code I've written in a long time -_-

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Krivodonova.cpp:
```diff
@@ -0,0 +1,2900 @@
+    package_data(order);
+  }
+  two_neighbors();
+  test_different_values_different_tensors();
```
For increased clarity, I suggest renaming these functions to
- `test_package_data`, to avoid any ambiguity with the function being tested
- `test_limiting_two_neighbors`
- `test_limiting_different_values_different_tensors`

Could also do similar renames in 2D,3D cases.

nilsdeppe replied:
Sure, but just to confirm, the function names will then be `test_1d::test_package_data`, etc. This is what you want?

fmahebert replied:
Yes, I agree it's slightly redundant, but it will make it clearer at the call site (or invoke as `test_1d::package_data`, to the same effect).

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Krivodonova.cpp:
```diff
@@ -0,0 +1,2900 @@
+  // [(0,0), (1,0), (2,0), (0,1), (1,1), (2,1), (0,2), (1,2), (2,2)]
+  //  (0,     1,     2,     3,     4,     5,     6,     7,     8)
+
+  // Apply to (2,2)
```
Could you document a little more verbosely what is happening here? Is each of these `test_X_Y` functions testing all the different ways in which the `(X,Y)` coeff can be modified? If so perhaps add a statement saying roughly that. Consider renaming to `test_limiting_of_X_Y_coeff`, or similar.

fmahebert reviewed (commented):
I think we are converging here 👍. Here are a few more suggestions for documentation clarification and some possible code cleanups.

nilsdeppe commented:
@fmahebert @nilsleiffischer I've pushed a fixup commit addressing your comments.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Krivodonova.hpp:
```diff
@@ -620,8 +615,8 @@ bool Krivodonova<VolumeDim, tmpl::list<Tags...>>::operator()(
         "collocation points, bases, and quadrature in each direction. The "
-        "extents are: "
-        << mesh.extents());
+        "meshes are: "
```
"meshes are" -> "mesh is"

fmahebert reviewed (commented):
LGTM 👍. I am happy for you to squash the fixup (including this tiny mistake) once @nilsleiffischer is also ready.

nilsvu commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Krivodonova.hpp:
```diff
@@ -0,0 +1,995 @@
+  explicit Krivodonova(
+      std::array<double,
+                 Spectral::maximum_number_of_points<Spectral::Basis::Legendre>>
+          alphas,
```
You could default this argument to `Alphas::default_value()` if you like (making that a `constexpr`)

nilsdeppe replied:
I agree with this reducing text but I don't agree with the conceptual connection it builds. I view the options as being a layer on top of the core class and so the class shouldn't rely on the options :)

nilsvu reviewed (commented):
Yes please squash, including this if you choose to do it

nilsdeppe commented:
@fmahebert @nilsleiffischer thanks for the feedback! Rebased on develop and squashed :) @fmahebert I included the wording change you requested and squashed it.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Krivodonova.hpp:
```diff
@@ -0,0 +1,995 @@
+        // times we call minmod, not because it is required for correctness.
+        if (UNLIKELY(i == j)) {
+          const double tmp = minmod(i, j, tensor_storage_index);
+          if (tmp == self_coeffs[mesh.storage_index(Index<VolumeDim>{i, j})]) {
```
Two ideas that could make the code look a bit cleaner, but are unlikely to have a measurable performance impact (so I am writing them here for the record, but not as a change request):
- You call `mesh.storage_index(i, j)` several times per (i, j) pair (including inside the `minmod` lambda). There's an opportunity to precompute and store the mesh storage index.
- You do a lot of `double == double` checks in this algorithm. In my Minmod limiter I opted to have the minmod function return its result as a pair of `double, bool` where the `bool` denotes whether the `double` is the first argument of minmod. Indeed, the minmod function knows (or can trivially know) which argument it is returning so it seems silly to have to compute (after the fact) whether this argument was the first argument.

nilsdeppe replied:
I'll make an issue. Given how slow the load time of this PR is already, I'd rather get it merged :)

nilsdeppe commented:
@fmahebert @nilsleiffischer I've rebased on develop

kidder commented:
Current status:

- [x] @kidder needs to do cursory final review

kidder commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Krivodonova.hpp:
```diff
@@ -0,0 +1,995 @@
+ *
+ * where \f$\left\{\xi, \eta, \zeta\right\}\f$ are the logical coordinates,
+ * \f$P_{i}\f$ are the Legendre polynomials, and the superscript \f$\{l,m,n\}\f$
+ * represents the element indexed by \f$l,m,n\f$. The coefficients are limited
```
define N_i

kidder commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Krivodonova.hpp:
```diff
@@ -0,0 +1,995 @@
+ * \f$c_{N,N,N}\f$, followed by \f$c_{N,N,N-1}\f$, \f$c_{N,N-1,N}\f$, and
+ * \f$c_{N-1,N,N}\f$. A detailed example is given below. Limiting is stopped if
+ * all symmetric pairs of coefficients are left unchanged, i.e.
+ * \f$c_{i,j,}=\tilde{c}_{i,j,k}\f$. By all symmetric coefficients we mean that,
```
missing `k` after `i,j,`

kidder commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Krivodonova.hpp:
```diff
@@ -0,0 +1,995 @@
+ *          \alpha_1\left(c^{l,m,n}_{0,0,2}-c^{l,m-1,n}_{0,0,2}\right),
+ *   \right. \\
+ *   &\;\;\;\;\left.
+ *          \alpha_2\left(c^{l,m,n+1}_{0,0,1}-c^{l,m,n}_{0,0,1}\right),
```
`{0,0,1}` should be `{0,1,1}` four times

kidder reviewed (changes requested):
you can squash these

kidder commented:
@fmahebert @nilsleiffischer this needs re-approval after minor documentation fix-ups
